### PR TITLE
Custom DreamWorld Support

### LINKED ISSUE #916: custom simulations in other systems
### Feature

- [ ] the ability to create custom simulation spawns and the ability to create new ones in other systems (maybe even an inception feature?). as of right now new horizons cannot let you enter the simulation outside the solar system (and if it does, it's really buggy).

### Context

this feature could allow for creation of whole different worlds inside worlds, and a great enhancement for some dlc-based mods that wish to use it.

splitting these issue into parts for the convenience of the nh dev team. also i am aware the simulation creation features are very difficult to implement and i do not expect them to be done soon, if not at all. i just hope to see it in new horizons someday.

### PR DESCRIPTION
**Pull request by Hawkbat**
## Major features
- Added `Dream.inDreamWorld` to planet configs to treat a planet as if it were part of the dream world dimension. The planet's contents will be deactivated unless you enter via a dream arrival point or another dream world dimension. Closes #916
- Added `dreamCampfires` and `dreamArrivalPoints` as new linked prop types, which allow you to create new entry points into Echoes of the Eye's dream world or create similar experiences using custom planets.

## Improvements
- General support for dream world simulations even if the vanilla dream world is disabled in custom solar systems.

### CODE REVIEW

File: NewHorizons/Builder/Props/EchoesOfTheEye/DreamCampfireBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            campfire._dreamArrivalLocation = DreamHandler.GetDreamArrivalLocation(info.id);
+
+            Delay.FireInNUpdates(() => {
+                var streaming = campfireObj.GetComponentInChildren<DreamCampfireStreaming>();
+                if (streaming != null)
+                {
+                    var targetArrivalPoint = Locator.GetDreamArrivalPoint(campfire._dreamArrivalLocation);
+                    if (targetArrivalPoint != null)
+                    {
+                        var streamingGroup = targetArrivalPoint.transform.root.GetComponentInChildren<StreamingGroup>();
```
Comment (JohnCorby) [inline NewHorizons/Builder/Props/EchoesOfTheEye/DreamCampfireBuilder.cs:59]: use the StreamingHandler method to get this

File: NewHorizons/Builder/Props/EchoesOfTheEye/DreamSimulationBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            if (materials.Any(m => m.name.Contains("Ocean_Stencil_mat"))) return waterMaterial;
+            return gridMaterial;
+        }
+
+        private static bool CheckMeshCreationHeuristic(GameObject go, Material[] materials)
+        {
+            if (go.layer == Layer.DreamSimulation) return false;
+            var mr = go.GetComponent<MeshRenderer>();
+            if (EXCLUDED_SHADER_NAMES.Any(name => materials.Any(mat => mat.shader.name.Contains(name)))) return false;
+            if (go.transform.parent)
```
Comment (JohnCorby) [inline NewHorizons/Builder/Props/EchoesOfTheEye/DreamSimulationBuilder.cs:60]: parent layer do not effect rendering, right? so should be able to check just this layer

File: NewHorizons/Patches/EchoesOfTheEyePatches/DreamworldControllerPatches.cs [hunk 3]
```diff
@@ -26,5 +26,40 @@ public static IEnumerable<CodeInstruction> DreamWorldController_SpawnInDreamworl
                 .Repeat(matcher => matcher.RemoveInstructions(5))
                 .InstructionEnumeration();
         }
+
+        [HarmonyPrefix]
+        [HarmonyPatch(nameof(DreamWorldController.EnterDreamWorld))]
+        public static void DreamWorldController_EnterDreamWorld(DreamWorldController __instance, DreamCampfire dreamCampfire, DreamArrivalPoint arrivalPoint)
+        {
+            var dreamWorldAO = Locator.GetAstroObject(AstroObject.Name.DreamWorld);
```
Comment (JohnCorby) [inline NewHorizons/Patches/EchoesOfTheEyePatches/DreamworldControllerPatches.cs:34]: maybe comment this, idc

File: NewHorizons/Builder/General/AmbientLightBuilder.cs [hunk 4]
```diff
@@ -64,6 +65,8 @@ public static Light Make(GameObject planetGO, Sector sector, AmbientLightModule
                 }
             }
 
+            lightGO.AddComponent<AmbientLight>();
```
Comment (JohnCorby) [inline NewHorizons/Builder/General/AmbientLightBuilder.cs:68]: this is really goofy. dw should support ambient lights

**Review by JohnCorby — COMMENTED** [review]
lookin good tho, theres just some extra stuff to do like not changing far clip plane and layers and stuff i think

oh and entryway bullshit i think. check the discord convo

File: NewHorizons/Builder/Props/EchoesOfTheEye/DreamSimulationBuilder.cs (hunk 2 above)
Comment (Hawkbat) [reply NewHorizons/Builder/Props/EchoesOfTheEye/DreamSimulationBuilder.cs:60]: Needed this to handle props like Prefab_IP_DreamFire where the existing non-simulation renderers are siblings of the simulation renderers in the same props. Don't want to create simulation meshes if the prop already contains them. The heuristic runs on every object with a mesh renderer so the parent object not matching the heuristic won't prevent the child objects from getting simulation meshes.

**Review by xen-42 — CHANGES_REQUESTED** [review]
We should also add a `inDreamworld` field to planets that makes them invisible until you enter the dreamworld and makes them invisible again on exit. 

If we disable all children of a planets root object except for `FieldDetector` it should keep orbiting while disabled. 

Only disable a few frames after start.

File: NewHorizons/Builder/Props/EchoesOfTheEye/DreamCampfireBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                        if (streamingGroup)
+                        {
+                            streaming._streamingGroup = streamingGroup;
+                        }
+                    }
+                }
+            }, 2);
```
Comment (xen-42) [inline NewHorizons/Builder/Props/EchoesOfTheEye/DreamCampfireBuilder.cs:66]: Ought to leave a comment explaining why 2, even if its just "idk i didnt work at 1 so i did 2"

File: NewHorizons/Components/EOTE/DreamDimension.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+        void UpdateState()
+        {
+            foreach (var obj in toggledObjects) obj.SetActive(active);
+        }
+
+        public void Update()
+        {
+            if (!initialized) return;
+            SetActive(PlayerState.InDreamWorld());
```
Comment (xen-42) [inline NewHorizons/Components/EOTE/DreamDimension.cs:45]: Is it not better to use the entered/exited dreamworld events? If theres a reason we cant do that we should comment that here

**Review by xen-42 — CHANGES_REQUESTED** [review]

File: NewHorizons/Components/EOTE/DreamDimension.cs (hunk 6 above)
Comment (Hawkbat) [reply NewHorizons/Components/EOTE/DreamDimension.cs:45]: Nah you're right, should just be using the globalmessenger events.

**Review by xen-42 — CHANGES_REQUESTED** [review]
When dying, we ought to preload streaming stuff of where we're about to wake up. Right now waking up on TH in Examples is very obvious.

Weird artifact when sleeping in examples system, only shows when looking at certain angles
image

Do NH-made water volumes extinguish lanterns?

**Review by xen-42 — APPROVED** [review]

**Pull request merged**